### ISSUE TITLE
Find a way to let users place orders

### ISSUE THREAD

**Issue by Folas1337** [Meta-Union/Meta-Union#20]
The general idea is this:
> User1 has a demand like "I want 100 stone and I will pay 100 MetaCoins"
Currently there is no way other than asking in game for someone to help you out with that, which does not help when nobody is online or people don't answer for whatever reason.

So the idea is for a kind of "virtual blackboard" or any other way to place these orders.
Basically user1 can set up an order for the 100 stone and whoever delivers it in whatever way, gets paid 100 MetaCoins.

**Comment by Folas1337** [comment]
A plugin called [Companies](https://www.spigotmc.org/resources/companies.109558/) is actually looking into this and said this could be a priority and become a reality somewhat soon, see also their issue https://github.com/eldoriarpg/companies/issues/156

I'll be asking around some more to see if anyone's interested in developing this and we'll see who comes out on top :D

**Comment by Folas1337** [comment]
ArasBuilds, as much as he might hate me to say this, is also interested in developing a plugin for this very use case.

We'll see who wins this race and which solution convinces me most.
Aras, if you are reading this/ever read this, this is only for me to keep track, sorry for "outing" you <3

**Comment by MRollin03** [comment]
Yeah yeah. tbh, I see the already developed plugin win the race. but I'll learn for future plugins that does not exist :)

**Comment by Folas1337** [comment]
> Yeah yeah. tbh, I see the already developed plugin win the race. but I'll learn for future plugins that does not exist :)

I might switch to yours eventually, perhaps ;P
As long as all you need to be happy is a good exercise, this shall suffice <3

**Comment by Folas1337** [comment]
Current situation: The Company plugin has turned out to have some pretty bad design choices which makes it unfeasible to use on Meta-Union. As it turns out, most of it relies on a chat GUI, which means you have to click a bunch of things in the chat but Bedrock Edition can't do that so they'd be left behind.

Du to that we will not be getting the company plugin but instead Aras is going to see if he can make a plugin himself, his current estimation was that it will take about 2 months until we have a finished plugin.

Here's a quick overview of what the request is:
- `/order create <item|hand> <amount> <price>` This command will let you place an order for either an item you specify or the item you are holding in your hand. You also specify the amount you request and the price you are going to pay.

Here are a few examples of how it could be used:
- `/order create oak_planks 128 100` This would put up an order for 128 oak planks at the price of 100 MetaCoins.
- `/order create hand 64 1000` If you held diamonds while using this command, you would be ordering a stack of diamonds and pay 1000 MetaCoins.
- `/order create nether_star 1 100` would ask for 1 nether star at the price of 100 MetaCoins.

You can list orders by using `/order list [user|item]`. Using just `/order list` will open a Chest GUI where you see all the orders. Using `/order list Folas1337` would show only the orders Folas1337 has put up. Using `/order list diamond` would show any orders that are asking for diamonds.

Through this list you can click on an order, then be asked to confirm with another click which also tells you that you are going to pay if you have the required material. It then checks your inventory if you have the items it requires and also if you have enough money (take into account the fee from the config!) Then the items are taken from your inventory, the money is charged and the items are put into a virtual chest of the player who ordered it.

It should also be possible to cancel the order from the `/order list [user|item]` menu. So if I wanted to cancel one of my orders, I could use `/order list Folas1337`, it would show my orders and by clicking on one of them, it would allow me (and only me) to cancel it with another click on a cancel button. Obviously this should only be possible if you are the owner of the order OR if you have permission to do so.

And lastly to take out items you have received through the order, we'll also be having a chest GUI which you open with `/order received` and by clicking on the orders, it puts the items right in your inventory or does not if you do not have enough space in your inventory.

List of permissions:
- orderplugin.create to be able to create orders
- orderplugin.list to be able to use the `/order list [user|item]` command
- orderplugin.cancel.self to be able to cancel your own orders
- orderplugin.cancel.others to be able to cancel other people's orders
- orderplugin.received to be able to use `/order received`

I imagine the config.yml to have the following options:
```yaml
order-fee: 10 # The fee the person who fulfills the order will be paying. This is mainly to prevent using orders as an infinite supply of items
order-fee-percentage: true #If this is set to true, the fee is a percentage of the initial price rather than a fixed amount of money
order-limit: 5 #The maximum amount of orders you can place at once
order-limit-item-count: 320 #This is the maximum amount of items you can ask for per order. 320 = 5 stacks. This limits the amount in '/order create <item|hand> <amount> <price>`
order-timeout: 10080 #The timeout for an order in minutes. This example means 7 days.
```

So let's go through an entire example, assuming the above config.yml:
- `/order create diamond 1000 10` You wouldn't be able to use this command as it asks for more than 320 items
- `/order create diamond 64 1000` This would create an order for 64 diamonds and whoever fulfills it, gets 1000 MetaCoins.
- Afterwards someone uses `/order list diamond` which also shows the order created above but the price will show up as 1100 MetaCoins as the 10% fee is applied. Clicking on it, will open the chest GUI and ask for confirmation. Once confirmed, it turns out you do not have 64 diamonds in your inventory, so it is cancelled.
- Someone else now tries to accept the order but has exactly 1000 MetaCoins in their account. It would then be denied as you can't pay 1100 MetaCoins.
- Yet another person tries to accept it as well, has 128 diamonds in their inventory as well as 2000 MetaCoins in their account. It then removes 64 diamonds from their inventory and takes 1100 MetaCoins from the account, leaving them with 900 MetaCoins.
- If nobody accepted it within 7 days, the order would be removed from the list
- If the creator of the order used `/order list` and cancelled the order, it would be removed from the `/order list` GUI. Also make sure that if people have the GUI open and then accept an order, it is not processed but instead the user is told that the order is no longer available
- If the order was fulfilled, the creator could use `/order received` and take out the items as described

**Comment by MRollin03** [comment]
The current repository for the plugin: https://github.com/MRollin03/JobActions
Testing in terms of exploits and bugs is required for further state implementation validation.

**Comment by Folas1337** [comment]
I'll try to get around to this as soon as possible, I'll comment whenever I've done anything regarding it :)

**Comment by Folas1337** [comment]
I found a few issues which have been raised as such on the JobActions repo. I'll wait for the fixes to come around soon and then hopefully we'll have a high quality plugin! <3

**Comment by Folas1337** [comment]
I'm declaring this as blocked once again until the last couple bugs in the plugin are sorted :)
We're constantly getting closer to adding the plugin however.

**Comment by Folas1337** [comment]
A plugin to potentially consider: https://modrinth.com/plugin/karta-playercontract

**Comment by Folas1337** [comment]
As much as I used to love this idea, I don't think it's really viable anymore.

It would **technically** still be awesome to have but adding yet another way to buy and sell items seems to get too convoluted. Meta-Union already has a Server Shop with many different categories and an auction house to sell items in. Also there are chest shops where you can sell things. Having another system where you can request items (which the chest shops technically can do as well) seems like too much at this point.

@MRollin03 I'm sorry I potentially caused you some work while never actually making use of it, I hope you can forgive me :yellow_heart:

Due to that I'll close the issue 👋
